lifecycle-operator/controllers/common/eventsender.go

```diff
@@ -15,28 +20,100 @@ type IEvent interface {
-	return newK8sSender(recorder)
+type EventMultiplexer struct {
+	logger   logr.Logger
+	emitters []IEvent
+}
+
+func NewEventMultiplexer(logger logr.Logger, recorder record.EventRecorder, client ce.Client) *EventMultiplexer {
+	multiplexer := &EventMultiplexer{
+		logger: logger,
+	}
+	multiplexer.register(newCloudEventSender(logger, client))
+	multiplexer.register(NewK8sSender(recorder))
+	return multiplexer
+}
+
+func (e *EventMultiplexer) register(emitter IEvent) {
+	if emitter != nil {
+		e.emitters = append(e.emitters, emitter)
+	}
+}
+
+func (e *EventMultiplexer) Emit(phase apicommon.KeptnPhaseType, eventType string, reconcileObject client.Object, status string, message string, version string) {
+	for _, emitter := range e.emitters {
+		e.logger.Info(fmt.Sprintf("Emitting event using %T", emitter))
+		emitter.Emit(phase, eventType, reconcileObject, status, message, version)
+	}
 }
 
 // ===== Cloud Event Sender =====
-// TODO: implement Cloud Event logic
+
+type cloudEvent struct {
+	client ce.Client
+	logger logr.Logger
+}
+
+func newCloudEventSender(logger logr.Logger, client ce.Client) *cloudEvent {
+	return &cloudEvent{
+		client: client,
+		logger: logger,
+	}
+}
+
+// Emit creates a Cloud Event and send it to the endpoint
+func (e *cloudEvent) Emit(phase apicommon.KeptnPhaseType, eventType string, reconcileObject client.Object, status string, message string, version string) {
+	endpoint := config.Instance().GetCloudEventsEndpoint()
+	if endpoint == "" {
+		// if no endpoint is configured we don't emit any event
+		if !strings.HasPrefix(endpoint, "http") {
+			e.logger.V(5).Info(fmt.Sprintf("CloudEvent endpoint configured but it does not start with http: %s", endpoint))
+		}
+		return
+	}
+	event := ce.NewEvent()
+	event.SetSource("keptn.sh")
+	event.SetType(fmt.Sprintf("%s.%s", phase.LongName, status))
+
+	msg := setEventMessage(phase, reconcileObject, message, version)
+	err := event.SetData(ce.ApplicationJSON, map[string]interface{}{
+		"message": msg,
+		"type":    eventType,
+		"version": version,
+		"resource": map[string]string{
+			"group":     reconcileObject.GetObjectKind().GroupVersionKind().Group,
+			"kind":      reconcileObject.GetObjectKind().GroupVersionKind().Kind,
+			"version":   reconcileObject.GetObjectKind().GroupVersionKind().Version,
+			"name":      reconcileObject.GetName(),
+			"namespace": reconcileObject.GetNamespace(),
+		},
+	})
+	if err != nil {
+		e.logger.V(5).Info(fmt.Sprintf("Failed to set data for CloudEvent: %v", err))
+		return
+	}
+
+	ctx := ce.ContextWithTarget(context.TODO(), endpoint)
+	if result := e.client.Send(ctx, event); ce.IsUndelivered(result) {
+		e.logger.V(5).Info(fmt.Sprintf("Failed to send CloudEvent: %v", event))
+	}
+}
 
 // ===== K8s Event Sender =====
 
 type k8sEvent struct {
 	recorder record.EventRecorder
 }
 
-func newK8sSender(recorder record.EventRecorder) IEvent {
+func NewK8sSender(recorder record.EventRecorder) IEvent {
```

Comment:
if we have `newCloudEventSender`, let's be consistent in the naming
```suggestion
func NewK8sEventSender(recorder record.EventRecorder) IEvent {
```

Reply:
CloudEvent is the name of the technology itself 😅

Reply:
Yeah but `K8sSender` sounds a little bit weird :D but I won't push for naming as a blocker :)